File: ejercicios_simpy/simulation_9.py
import simpy
import random
import statistics
from collections import defaultdict

class SimuladorSecuencial:
    def __init__(self, env, mean_interarrival, mean_service, 
                 servers_level1, servers_level2, distribution_type):
        self.env = env
        self.mean_interarrival = mean_interarrival
        self.mean_service = mean_service
        self.distribution_type = distribution_type
        self.server1 = simpy.Resource(env, capacity=servers_level1)
        self.server2 = simpy.Resource(env, capacity=servers_level2)
        self.queue1_delays = []
        self.queue2_delays = []
        self.system_times = []
        self.queue1_lengths = []
        self.queue2_lengths = []
        self.server1_busy = []
        self.server2_busy = []
        self.num_completed = 0
        self.num_arrivals = 0
        self.queue1_entry_times = {}
        self.queue2_entry_times = {}
        self.env.process(self.customer_generator())
        self.env.process(self.monitor_queues())
        self.env.process(self.monitor_servers())
    
    def get_interarrival_time(self):
        if self.distribution_type in [1, 3]:
            return random.expovariate(1.0 / self.mean_interarrival)
        else:
            return self.mean_interarrival
    
    def get_service_time(self):
        if self.distribution_type in [1, 2]:
            return random.expovariate(1.0 / self.mean_service)
        else:
            return self.mean_service
    
    def customer_generator(self):
        while True:
            yield self.env.timeout(self.get_interarrival_time())
            
            arrival_time = self.env.now
            self.num_arrivals += 1
            customer_id = self.num_arrivals
            self.env.process(self.customer_flow(customer_id, arrival_time))
    
    def customer_flow(self, customer_id, arrival_time):
        queue1_entry = self.env.now
        self.queue1_entry_times[customer_id] = queue1_entry
        
        with self.server1.request() as req:
            yield req
            
            queue1_delay = self.env.now - queue1_entry
            self.queue1_delays.append(queue1_delay)
            del self.queue1_entry_times[customer_id]
            yield self.env.timeout(self.get_service_time())
        
        queue2_entry = self.env.now
        self.queue2_entry_times[customer_id] = queue2_entry
        with self.server2.request() as req:
            yield req
            
            queue2_delay = self.env.now - queue2_entry
            self.queue2_delays.append(queue2_delay)
            del self.queue2_entry_times[customer_id]
            
            yield self.env.timeout(self.get_service_time())
        
        system_time = self.env.now - arrival_time
        self.system_times.append(system_time)
        self.num_completed += 1
    
    def monitor_queues(self):
        while True:
            q1_len = len(self.server1.queue)
            self.queue1_lengths.append(q1_len)
            
            q2_len = len(self.server2.queue)
            self.queue2_lengths.append(q2_len)
            
            yield self.env.timeout(0.1) 
    
    def monitor_servers(self):
        while True:
            self.server1_busy.append(len(self.server1.users))
            self.server2_busy.append(len(self.server2.users))
            yield self.env.timeout(0.1)  
    
    def generate_report(self):
        def safe_mean(data):
            return statistics.mean(data) if data else 0.0
        
        def safe_max(data):
            return max(data) if data else 0.0
        
        def safe_min(data):
            return min(data) if data else 0.0
        
        report = "Multi-server queueing system using SimPy\n\n"
        report += f"Mean interarrival time{11*' '}{self.mean_interarrival:.3f} minutes\n\n"
        report += f"Mean service time{16*' '}{self.mean_service:.3f} minutes\n\n"
        report += f"Total simulation time{12*' '}{self.env.now:.3f} minutes\n\n"
        report += f"Distribution type: {self.distribution_type}\n"
        
        dist_map = {
            1: "Interarrival: Exponential, Service: Exponential",
            2: "Interarrival: Constant, Service: Exponential",
            3: "Interarrival: Exponential, Service: Constant",
            4: "Interarrival: Constant, Service: Constant"
        }
        report += dist_map.get(self.distribution_type, "Invalid distribution type") + "\n"
        report += f"Number of servers at level 1: {self.server1.capacity}\n"
        report += f"Number of servers at level 2: {self.server2.capacity}\n\n"
        
        report += "\nDelays in Queue 1, in minutes:\n"
        report += f"  Average: {safe_mean(self.queue1_delays):.3f}\n"
        report += f"  Maximum: {safe_max(self.queue1_delays):.3f}\n"
        report += f"  Minimum: {safe_min(self.queue1_delays):.3f}\n"
        report += f"  Number measured: {len(self.queue1_delays)}\n"
        
        report += "\nDelays in Queue 2, in minutes:\n"
        report += f"  Average: {safe_mean(self.queue2_delays):.3f}\n"
        report += f"  Maximum: {safe_max(self.queue2_delays):.3f}\n"
        report += f"  Minimum: {safe_min(self.queue2_delays):.3f}\n"
        report += f"  Number measured: {len(self.queue2_delays)}\n"
        
        report += "\nTotal System Time (arrival to final departure), in minutes:\n"
        report += f"  Average: {safe_mean(self.system_times):.3f}\n"
        report += f"  Maximum: {safe_max(self.system_times):.3f}\n"
        report += f"  Minimum: {safe_min(self.system_times):.3f}\n"
        report += f"  Number measured: {len(self.system_times)}\n"
        
        report += "\nQueue lengths:\n"
        report += f"  Average queue 1 length: {safe_mean(self.queue1_lengths):.3f}\n"
        report += f"  Average queue 2 length: {safe_mean(self.queue2_lengths):.3f}\n"
        
        server1_util = sum(self.server1_busy) / len(self.server1_busy) / self.server1.capacity * 100
        server2_util = sum(self.server2_busy) / len(self.server2_busy) / self.server2.capacity * 100
        report += "\nServer utilization:\n"
        report += f"  Server 1 utilization: {server1_util:.1f}%\n"
        report += f"  Server 2 utilization: {server2_util:.1f}%\n"
        
        report += f"\nTime simulation ended:{12*' '}{self.env.now:.3f} minutes\n"
        report += f"Number of customers completed both services:{8*' '}{self.num_completed}\n"
        
        return report

def read_parameters(filename):
    with open(filename, 'r') as f:
        values = list(map(float, f.readline().strip().split()))
        
        mean_interarrival = values[0]
        mean_service = values[1]
        total_simulation_time = int(values[2])
        distribution_type = int(values[3])
        servers_level1 = int(values[4])
        servers_level2 = int(values[5])
        
        return (mean_interarrival, mean_service, total_simulation_time,
                distribution_type, servers_level1, servers_level2)

def run_simulation(input_file, output_prefix):
    try:
        params = read_parameters(input_file)
        mean_interarrival, mean_service, total_simulation_time, \
        distribution_type, servers_level1, servers_level2 = params
        
        env = simpy.Environment()
        system = SimuladorSecuencial(env, mean_interarrival, mean_service,
                                   servers_level1, servers_level2, distribution_type)
        env.run(until=total_simulation_time)
        
        report = system.generate_report()
        output_filename = f"{output_prefix}_{distribution_type}.txt"
        
        with open(output_filename, 'w') as f:
            f.write(report)
        
        print(f"Simulation complete. Report saved to {output_filename}")
        return True
    
    except FileNotFoundError:
        print(f"Error: Cannot open input file {input_file}")
        return False
    except Exception as e:
        print(f"Error running simulation: {str(e)}")
        return False

if __name__ == "__main__":
    input_file = "parameters_9.txt"
    output_prefix = "report_9"
    
    if not run_simulation(input_file, output_prefix):
        print("Creating sample parameters file and running simulation...")
        with open(input_file, 'w') as f:
            f.write("5.0 4.0 1000 1 2 3\n")
        
        run_simulation(input_file, output_prefix)

File: notebooks/simulation_9.c

#include "simlib.h"

#define EVENT_ARRIVAL 1
#define EVENT_DEPARTURE1 2
#define EVENT_DEPARTURE2 3
#define LIST_QUEUE1 1
#define LIST_SERVER1 2
#define LIST_QUEUE2 3
#define LIST_SERVER2 4
#define SAMPST_DELAYS_Q1 1    
#define SAMPST_DELAYS_Q2 2    
#define SAMPST_TOTAL_TIME 3  
#define STREAM_INTERARRIVAL 1
#define STREAM_SERVICE 2

int num_custs_delayed, total_simulation_time, num_completed_customers, distribution_type;
int servers_level1, servers_level2;  
float mean_interarrival, mean_service;
FILE *infile, *outfile;

float get_interarrival_time() {
    switch (distribution_type) {
        case 1: 
        case 3:
            return expon(mean_interarrival, STREAM_INTERARRIVAL);
        case 2: 
        case 4:
            return mean_interarrival;
        default:
            return expon(mean_interarrival, STREAM_INTERARRIVAL);
    }
}

float get_service_time() {
    switch (distribution_type) {
        case 1: 
        case 2:
            return expon(mean_service, STREAM_SERVICE);
        case 3: 
        case 4:
            return mean_service; 
        default:
            return expon(mean_service, STREAM_SERVICE); 
    }
}

void init_model(void) {
    num_custs_delayed = 0;
    num_completed_customers = 0;
    event_schedule(sim_time + get_interarrival_time(), EVENT_ARRIVAL);
}

void arrive(void) {
    event_schedule(sim_time + get_interarrival_time(), EVENT_ARRIVAL);

    transfer[1] = sim_time;  
    transfer[2] = sim_time;  

    if (list_size[LIST_SERVER1] >= servers_level1) {
        list_file(LAST, LIST_QUEUE1);
    }
    else {
        sampst(0.0, SAMPST_DELAYS_Q1);  
        ++num_custs_delayed;
        list_file(FIRST, LIST_SERVER1);
        event_schedule(sim_time + get_service_time(), EVENT_DEPARTURE1);
    }
}

void depart1(void) {
    list_remove(FIRST, LIST_SERVER1);

    transfer[3] = sim_time;

    if (list_size[LIST_SERVER2] >= servers_level2) {
        transfer[4] = sim_time; 
        list_file(LAST, LIST_QUEUE2);
    }
    else {
        sampst(0.0, SAMPST_DELAYS_Q2);
        list_file(FIRST, LIST_SERVER2);
        event_schedule(sim_time + get_service_time(), EVENT_DEPARTURE2);
    }

    if (list_size[LIST_QUEUE1] > 0) {
        list_remove(FIRST, LIST_QUEUE1);
        sampst(sim_time - transfer[2], SAMPST_DELAYS_Q1);  
        ++num_custs_delayed;
        list_file(FIRST, LIST_SERVER1);
        event_schedule(sim_time + get_service_time(), EVENT_DEPARTURE1);
    }
}

void depart2(void) {
    list_remove(FIRST, LIST_SERVER2);

    float system_time = sim_time - transfer[1];
    sampst(system_time, SAMPST_TOTAL_TIME);
    ++num_completed_customers;

    if (list_size[LIST_QUEUE2] > 0) {
        list_remove(FIRST, LIST_QUEUE2);
        sampst(sim_time - transfer[4], SAMPST_DELAYS_Q2);  
        list_file(FIRST, LIST_SERVER2);
        event_schedule(sim_time + get_service_time(), EVENT_DEPARTURE2);
    }
}

void report(void) {
    fprintf(outfile, "\nDelays in Queue 1, in minutes:\n");
    out_sampst(outfile, SAMPST_DELAYS_Q1, SAMPST_DELAYS_Q1);

    fprintf(outfile, "\nDelays in Queue 2, in minutes:\n");
    out_sampst(outfile, SAMPST_DELAYS_Q2, SAMPST_DELAYS_Q2);

    fprintf(outfile, "\nTotal System Time (arrival to final departure), in minutes:\n");
    out_sampst(outfile, SAMPST_TOTAL_TIME, SAMPST_TOTAL_TIME);

    fprintf(outfile, "\nQueue lengths and server utilization:\n");
    out_filest(outfile, LIST_QUEUE1, LIST_SERVER1);
    out_filest(outfile, LIST_QUEUE2, LIST_SERVER2);

    fprintf(outfile, "\nTime simulation ended:%12.3f minutes\n", sim_time);
    fprintf(outfile, "Number of customers completed both services:%8d\n", num_completed_customers);
    fprintf(outfile, "Number of servers at level 1: %d\n", servers_level1);
    fprintf(outfile, "Number of servers at level 2: %d\n\n", servers_level2);
}

int main() {
    char outfilename[100];

    infile = fopen("parameters_9.txt", "r");
    if (infile == NULL) {
        printf("Error: cannot open input file parameters_9.txt\n");
        return 1;
    }

    fscanf(infile, "%f %f %d %d %d %d", 
           &mean_interarrival, &mean_service, 
           &total_simulation_time, &distribution_type,
           &servers_level1, &servers_level2);

    while (fgetc(infile) != '\n');
    if (fscanf(infile, " \"%99[^\"]\"", outfilename) != 1) {  
        fseek(infile, -1, SEEK_CUR); 
        if (fscanf(infile, "%99s", outfilename) != 1) {
            sprintf(outfilename, "report_9_%d.txt", distribution_type);
        }
    }

    outfile = fopen(outfilename, "w");
    if (outfile == NULL) {
        printf("Error: cannot open output file %s\n", outfilename);
        fclose(infile);
        return 1;
    }

    fprintf(outfile, "Multi-server queueing system using simlib\n\n");
    fprintf(outfile, "Mean interarrival time%11.3f minutes\n\n", mean_interarrival);
    fprintf(outfile, "Mean service time%16.3f minutes\n\n", mean_service);
    fprintf(outfile, "Total simulation time%12d minutes\n\n", total_simulation_time);
    fprintf(outfile, "Distribution type: %d\n", distribution_type);
    switch (distribution_type) {
        case 1:
            fprintf(outfile, "Interarrival: Exponential, Service: Exponential\n");
            break;
        case 2:
            fprintf(outfile, "Interarrival: Constant, Service: Exponential\n");
            break;
        case 3:
            fprintf(outfile, "Interarrival: Exponential, Service: Constant\n");
            break;
        case 4:
            fprintf(outfile, "Interarrival: Constant, Service: Constant\n");
            break;
        default:
            fprintf(outfile, "Invalid distribution type. Defaulting to Exponential/Exponential.\n");
    }
    fprintf(outfile, "Number of servers at level 1: %d\n", servers_level1);
    fprintf(outfile, "Number of servers at level 2: %d\n\n", servers_level2);

    maxatr = 10;
    maxlist = 26;
    init_simlib();
    init_model();
    while (sim_time < total_simulation_time) {
        timing();
        switch (next_event_type) {
            case EVENT_ARRIVAL:
                arrive();
                break;
            case EVENT_DEPARTURE1:
                depart1();
                break;
            case EVENT_DEPARTURE2:
                depart2();
                break;
        }
    }
    report();
    fclose(infile);
    fclose(outfile);

    return 0;
}
